Playwright: add RestAPIClient class to interact with the WP REST API.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR adds the `RestAPIClient` class, used to interact with the WP REST API.

Key changes:
- update the SecretsManager validation scheme.
- add `RestAPIClient`, tasked with interacting with the WP REST API.

#### Testing instructions

1. pull the branch.
2. decrypt secrets.
3. edit a spec file to call `console.log(TestAccounts.getBearerToken())`

Depends on https://github.com/Automattic/wp-calypso/pull/63183.

worldomonation commented:
@dpasque I wanted to get your input at an early stage for the future design of ways to interact with the API.

Broadly speaking, I think there are two ways to implement the class for API calls.

### As part of TestAccounts object

This is the approach taken in the current implementation. 

**Advantages**
- only one object to manage when writing specs.
- as API calls are authenticated for a specific user, it makes intuitive sense that individual TestAccount makes API requests from the object.

**Disadvantages**
- as more method are implemented this can bloat the TestAccounts object.
- TestAccount object will end up doing multiple things, not just one.

### As a separate object

This approach would see a new object, say `APICalls` handle interaction with the API.

A rough outline may look like this:
```typescript
export class APICalls {
	readonly bearerToken: string
	readonly testAccount: TestAccount

	constructor( testAccount: TestAccount ) {
		this.testAccount = testAccount;
		this.bearerToken = testAccount.bearerToken;
	}

	// methods...
	async me() {}

	async deleteSite( siteName: string ) {}

	async makePost( content : string ) 
}
```

**Advantages**
- separation of concerns and less clutter of TestAccounts.

**Disadvantages**
- one more object the test writer has to instantiate and keep track of.
- where to put the initial method to obtain the `bearer token`.

worldomonation commented on packages/calypso-e2e/src/lib/test-account.ts:
```diff
@@ -10,15 +11,25 @@ import {
 import { TOTPClient } from '../totp-client';
 import { LoginPage } from './pages/login-page';
 
+interface BearerTokenResponse {
+	success: string;
+	data: {
+		bearer_token: string;
+		token_links: string[];
+	};
+}
+
```
The return type of the `.json()` method is [`unknown`](https://github.com/sindresorhus/got/blob/main/source/core/response.ts#L132) and so this interface is necessary.

It would be nice if I didn't have to do this though. I tried coercing the value as a JSON but doing so results in errors when I attempt to access the data in the structure:

`this.bearerToken = response.data.bearer_token;`.

dpasque replied:
Oh, I would say this is common practice actually! A lot of these HTTP request libraries don't type the responses (or some take generic types, so you tell it what type you expect the response to be). This looks great! 😄

worldomonation commented on yarn.lock:
```diff
@@ -244,6 +244,7 @@ __metadata:
     "@types/totp-generator": ^0.0.3
     "@wordpress/i18n": ^4.2.4
     asana-phrase: ^0.0.8
+    got: 11.8.3
```
`got` is now on v12.x.x but that drops CommonJS export, which causes problems when imported and used within `calypso-e2e`.

dpasque commented on packages/calypso-e2e/src/lib/test-account.ts:
```diff
@@ -1,6 +1,7 @@
 import fs from 'fs/promises';
 import path from 'path';
 import chalk from 'chalk';
+import got from 'got';
```
Ack, I feel so bad saying this, but since this package already has an HTTP request library (`node-fetch`), we _probably_ should use that one instead of adding a new one. Having two different HTTP request libraries in the same project feels a bit redundant and might be confusing for folks adding code later.  Are you okay switching over to `node-fetch`? It has a [good API](https://www.npmjs.com/package/node-fetch)! (It's just the fetch API!)

worldomonation replied:
Oh bummer...I'll rework this in `node-fetch`.

worldomonation replied:
Fixed in c572c75b88c3213cf885b21230bdf1d732f0da23.

dpasque commented on packages/calypso-e2e/src/secrets/secrets-manager.ts:
```diff
@@ -35,6 +35,10 @@ export interface Secrets {
 	storeSandboxCookieValue: string;
 	testCouponCode: string;
 	wpccAuthPath: string;
+	wpApiSecrets: {
```
Tiny nit, but would you be okay if we renamed this property to be a bit more specific to what it is? This is really the Calypso Oauth Application. Maybe we could call it `calypsoOauthApplication`?

`wpApiSecrets` could be interpreted as some kind of admin API key which might confuse folks!

worldomonation replied:
Yeah, that sounds good to me. I was at a bit of a loss as to what I should call it so the `wpApiSecrets` was more of a placeholder.

worldomonation replied:
Fixed in c572c75b88c3213cf885b21230bdf1d732f0da23.

dpasque commented:
Also, I love where this is going! This is looking great 😄

worldomonation committed:
- re-implement API interaction wtih node-fetch instead of got.
- implement RestAPIClient class with initial methods to interact with the API.

worldomonation commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -0,0 +1,63 @@
+import fetch, { RequestInit } from 'node-fetch';
+
+type GenericDict = { [ key: string ]: string | number };
+type RequestMethods = 'post' | 'get';
+type RequestParams = {
+	method: 'post' | 'get';
+	headers: GenericDict;
+	body?: GenericDict;
+};
```
Typing is not my strength and I struggled somewhat to come up with these type definitions.

I read up on Generics in TypeScript again but I'm not certain if this definition of `GenericDict` is in fact a generic type.

The essence of what I'm trying to do with generic dict type here is to not have to know the contents of the `body` parameter, which will wildly differ depending on the request being made. By defining a `GenericDict` type as above it permits type checks without having to explicitly define each field.

dpasque replied:
What you have works! But I think it differs a little bit from TypeScript generics, which is more about passing in different types almost as arguments to a function. The docs go into it a little bit more [here](https://www.typescriptlang.org/docs/handbook/2/generics.html).

The idea would be then that you could define whatever interfaces you wanted for the headers and body, and pass them into the `sendRequest` function:

```typescript
private async sendRequest<Header, Body>( url: string, params: RequestParams<Header, Body> )
```

worldomonation commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -0,0 +1,63 @@
+	 * @param {RequestParams} params Parameters for the request.
+	 */
+	private async sendRequest( url: string, params: RequestParams ): Promise< string > {
+		const response = await fetch( new URL( url, BASE_URL ), params as RequestInit );
```
`fetch` expects a `RequestInit` so I've casted the custom `RequestParameters` type here. Would it have made sense to drop the custom type and use `RequestInit` throughout?

worldomonation committed:
- more speficic type for response given by /me/sites
- implement method to obtain list of My Sites

dpasque commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -0,0 +1,91 @@
+	sites: Site[];
+}
+
+const BASE_URL = 'https://public-api.wordpress.com/rest/v1.1/';
```
As a heads up, the prefixes on this route can change depending on the API call and version! We have a v1.2 and v1.3, e.g.. Also, there's the .org rest API.

Maybe we should just have this be the host `https://public-api.wordpress.com`? And each route definition can handle specifying the needed version?

dpasque commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -0,0 +1,91 @@
+	 * @param {RequestParams} params Parameters for the request.
+	 */
+	private async sendRequest( endpoint: string, params: RequestParams ): Promise< any > {
+		// Trim the initial forward slash.
+		if ( endpoint.startsWith( '/' ) ) {
+			endpoint = endpoint.slice( 1 );
+		}
```
As a heads up, if you keep using the `URL` constructor below, you don't need to do this trim -- it handles it for you! 🎉

worldomonation replied:
In my experience, this is not the case:

```typescript
await fetch( new URL('/me', 'https://public-api.wordpress.com/rest/v1.1/')
```

results in 

```
{
  error: 'not_found',
  message: 'The specified path was not found. Please visit P2gHKz-8-p2 for valid paths.'
}
```

worldomonation replied:
Interesting, so if I have it set up as follows:

```
new URL('/me', 'https://public-api.wordpress.com/rest/v1.1')
```

The resulting `URL` object looks like this:

```
URL {
  href: 'https://public-api.wordpress.com/me',
  origin: 'https://public-api.wordpress.com',
  protocol: 'https:',
  username: '',
  password: '',
  host: 'public-api.wordpress.com',
  hostname: 'public-api.wordpress.com',
  port: '',
  pathname: '/me',
  search: '',
  searchParams: URLSearchParams {},
  hash: ''
}
```

This is obviously incorrect hence the error!

If I implement the changes you suggest in https://github.com/Automattic/wp-calypso/pull/63312#discussion_r866887864 and make the request methods specify the path starting from the hostname, then it works.

dpasque replied:
Yeah, I think based on the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL), they are expecting that "base" parameter to just be a host! Kind of confusing, IMHO! 🤷

worldomonation replied:
Yep, no problem - I've taken your advice to have smaller methods that build out individual parts of the request params and have made changes to the way I piece together URLs as part of that change.

dpasque commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -0,0 +1,91 @@
+	 * @param {GenericDict} [body] Request body.
+	 * @returns {RequestParams} Completed request parameter.
+	 */
+	private buildRequestParams( method: RequestMethods, body?: GenericDict ): RequestParams {
+		const params: RequestParams = {
+			method: method,
+			headers: {
+				Authorization: `Bearer ${ this.bearerToken }`,
+				'Content-Type': 'application/json',
+			},
+		};
+
+		// POST requests require a body.
+		if ( method === 'post' ) {
+			if ( ! body ) {
+				throw new Error( 'Cannot build a POST request with undefined body.' );
+			}
+			params[ 'body' ] = body;
+		}
+
+		return params;
+	}
+
+	/**
+	 * Sends the request to the endpoint.
+	 *
+	 * @param {string} endpoint Endpoint to send the request.
+	 * @param {RequestParams} params Parameters for the request.
+	 */
+	private async sendRequest( endpoint: string, params: RequestParams ): Promise< any > {
+		// Trim the initial forward slash.
+		if ( endpoint.startsWith( '/' ) ) {
+			endpoint = endpoint.slice( 1 );
+		}
```
Totally optional, but based on my experience on building similar API clients in the past, I'm not sure this is going to be the most useful abstraction over time.

Requests can vary a lot!  Some API requests may require specific headers for example. Also, POST requests may not necessarily be JSON -- they could Form or just URL-encoded search params. Accounting for all that variability in an abstraction like this over time can grow in a messy way sometimes. 

I've found in the past that what helps the most is to make really micro abstractions for small pieces (such as as a function to concat the host and route, or a function to get a "starting" header object with the auth token set). Then just have each public function (like `getMySites`) just build the request as needed. 

It makes each public function a little more verbose, but it makes it really obvious what that function is doing and how the web request is constructed.

Again, totally optional, but I wanted to throw out my perspective in case it's helpful! 😄

worldomonation replied:
Hmm. It's a lot to think about for sure.

I've found that the interactions with the WP API so far have followed a template of being JSON requests authorized using the bearer token. You're right that we might run into an odd call in the future that may be different. 

I'll look a bit more into what you said (to split things out into smaller builder methods).

dpasque replied:
Yeah, it may be overkill, and what you have totally works!

dpasque commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -0,0 +1,91 @@
+import fetch, { RequestInit } from 'node-fetch';
+
+type RequestMethods = 'post' | 'get';
+type RequestParams = {
+	method: 'post' | 'get';
+	headers: GenericDict;
+	body?: GenericDict;
+};
+type GenericDict = { [ key: string ]: string | number };
```
Building on my comment below about changing the abstraction level, one nice perk is you wouldn't need these very dynamic types.

Each function could just parameters for exactly what info it needs, and an interface for the response data that is returns.

dpasque commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -0,0 +1,91 @@
+			if ( ! body ) {
+				throw new Error( 'Cannot build a POST request with undefined body.' );
+			}
+			params[ 'body' ] = body;
```
Quick FYI -- I think the fetch API expects JSON bodies to be strings created from `JSON.stringify`

worldomonation replied:
Hmm, it seems to have been working fine without the `JSON.stringify` method call, but I'll make the change nonetheless. Thanks.

dpasque replied:
Hmm, maybe this a change they made to the node version? Or in a new version they no longer require that? I just remember this from prior experience and the docs.

worldomonation replied:
Perhaps - or it might be because I don't have POST methods with JSON implemented, that may be why!

dpasque reviewed (commented):
I really like the overall design you have going here! The changes look really nice! 🙌 

I added a few more design suggestions that might help. Let me know what you think!

worldomonation committed:
Revise the RestAPIClient class based on feedback.
- use smaller methods to build up the request params.
- remove use of GenericDict.
- add JSDOc where missing.

worldomonation committed:
Move bearer token method out of TestAccounts.

worldomonation commented:
I've moved the bearer token methods out of `TestAccount` class, instead preferring to have the `RestAPIClient` class handle that operation as well.

It turns out that I had forgotten the `TestAccounts` class only works with users that are defined in the secrets file. Users that we create in the test scripts (as part of a signup test, for example) are obviously not part of the secrets file. This is where the blocker arose and instead of making significant changes to TestAccounts, I felt it was preferable to go in this current direction.

Heads up for @dpasque.

dpasque commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -0,0 +1,157 @@
+	 * @returns {string} Authorization header in the requested scheme.
+	 * @throws {Error} If a scheme not yet implemented is requested.
+	 */
+	private getAuthorizationHeader( scheme: 'bearer' ): string {
```
What do you think of making this async, and then having it call getBearerToken if the cached token is null?

worldomonation committed:
Change the `getBearerToken` method to async and make it private.
